### ISSUE TITLE
PaymentMethodNonceCreatedListener doesn't call when using ThreeDSecureRequest.VERSION_2

### ISSUE THREAD

**Issue by vtolochk** [braintree/braintree_android#268]
### General information

* Environment: 
Sandbox

* Android Version and Device:
Tested with Android 7.0 and 9.0

* Braintree dependencies: 
com.braintreepayments.api:braintree:3.6.0
com.braintreepayments.api:three-d-secure:3.6.0

### Issue description

After calling ThreeDSecure.performVerification function app opens 3DS window where I enter verification code. Then it returns and doesn't call onPaymentMethodNonceCreated as the documentation says.

I manage to get nonce from onActivityResult via data.getParcelableExtra(EXTRA_THREE_D_SECURE_LOOKUP) but it seems that this nonce is incorrect.

I am using 4000000000001091 card number from https://developers.braintreepayments.com/guides/3d-secure/migration/android/v3
for 3DS V2.

In the sandbox, it shows me:
Status - Lookup enrolled
Gateway Rejected - 3D Secure reason
3DS Version - 2.1.0
Liability Shift Possible - true
Liability Shifted - false

It works with 3DS V1 and it works with iOS SDK 3DS V2.

Please help!

**Comment by vtolochk** [comment]
Also, I have my UI in js so I don't extend a BaseActivity like in example.

Here is my code if it can help.
https://github.com/vtolochk/react-native-braintree-xplat/blob/master/android/src/main/java/com/pw/droplet/braintree/Braintree.java

I am not a Java developer, so It may not so clear.
But I have a **setup** method that I call to init braintreeFragment and add listeners.
Also, I have a ** getNonceWithThreeDSecure** method that I call to open 3DS verification window

**Comment by vtolochk** [comment]
It seems that the Demo app works as expected. I will try to integrate the Demo project inside my app and see if it will work

**Comment by scannillo** [comment]
Casting to `AppCompatActivity` is not the cause of the issue you're seeing. Also, you don't need to extend any `BaseActivity` like we do in our demo app [here](https://github.com/braintree/braintree_android/blob/master/Demo/src/main/java/com/braintreepayments/demo/BaseActivity.java). We just do it because we want to re-use it's functionality in a variety of our demo features (ex: `VenmoActivity`, `GooglePaymentActivity`, `CardActivity`, etc).

I am still unable to replicate the behavior.

This might be worth looking at. In your code I see in some cases you reference `this.mBraintreeFragment` and in others `mBraintreeFragment`. Make sure you are only referencing **one** instance of `BraintreeFragment` throughout this file.

**Comment by vtolochk** [comment]
@scannillo Yes, thanks. I wasn't diving into that code so I had written a bit folly things...

About reference to the *mBraintreeFragment* I completely sure it's right, and I tried without this.

I am trying to integrate the Demo app and will comment here if it works.

**Comment by vtolochk** [comment]
`TLDR;`

So, the SDK works but maybe we can improve it.

I am not a Java developer, so I may have some incorrect terminology.
As I am new for native android code I don't know how to debug it professional way, so I debug it by Log.d().

Firstly, I decided that problem comes from SDK itself and removed dependencies from maven, downloaded source code from this repository and add it to my project.

It didn't help. The behavior was the same - **PaymentMethodNonceCreatedListener** didn't call.

I started debugging and dive into the source code. As I wrote above it works with 3DS v1 but not 3DS v2. I found these flows and added a ton of **Log.d()** :)

In the case of v1, it opens a 3DS window via something that I don't understand but it doesn't matter at all cause it works.

In the case of v2, it opens a 3DS window via **performCardinalAuthentication** function. It creates an Intent and calls **fragment.startActivityForResult(intent, BraintreeRequestCodes.THREE_D_SECURE)**. Then fragment waits for the response inside **onActivityResult** function. I found that **onActivityResult** function didn't trigger. I dive into stackoverflow to find the solution.

I didn't find anything that could help. But in one moment I realized that my top level 
**onActivityResult** function calls. I start thinking about it.

`A little introduction to my project`
It's a react-native app. I have an android directory that contains an android project. Basically it contains MainApplication and MainActivity files. So I have one file for the main activity. Other activities react handles for me.

`Let's continue`

So It doesn't call fragment **onActivityResult** function.

I got a tag for that fragment and called it manually.
 **getSupportFragmentManager().findFragmentByTag("com.braintreepayments.api.BraintreeFragment").onActivityResult(requestCode, resultCode, data);**

Oh lala! It started working!

But wait, it was too early to celebrate.

It indeed called **onActivityResult**. But with wrong *requestCode*. As I understood you are using different requests codes to call different activities for results.

So I dived into stack overflow one more time.
I found the following advice:

``
You are calling startActivityForResult() from your Fragment. When you do this, the requestCode is changed by the Activity that owns the Fragment.
``

So my main Activity owns the Fragment. Feel free to correct me if I am wrong.

That advice was with the solution - instead of calling **startActivityForResult** start calling ** getActivity().startActivityForResult**.

But I need to change source code of Braintree...
But I gave it a try
Inside *BraintreeFragment.java* file, inside **startActivityForResult** function I changed this call -> **super.startActivityForResult(intent, requestCode)** to this -> **getActivity().startActivityForResult(intent, requestCode)**.

AND IT STARTED WORKING AS EXPECTED!

So now I have a couple of possible solutions:

THE BAD:
Create a pull request with this change, you will approve it and maybe some other react-native developers will be happy with V2 as me after several days of debugging :)

THE GOOD:
You will give me advice on how to make it work without changing sources.

THE UGLY:
Fork the library, put it to my project, change that single line of code and we are done.
But I am not sure if it will work as expected in production.
At least I will need to change **cardinalEnvironment** to the production.
And developers that will support this project after me probably will want to kill me...

I hope you read everything.
I would like to get an opinion from more experienced java developers about this case.

Thank you!

**Comment by scannillo** [comment]
Glad you got your issue resolved. It seems that contacting out tech support team at https://help.braintreepayments.com will be your best bet for troubleshooting your integration. Also our developer guides https://developers.braintreepayments.com/guides/overview have a lot of information that could have helped you through this process.

**Comment by pcova** [comment]
@vtolochk had you make it work without changing the source code of Braintree?

**Comment by vtolochk** [comment]
@pcova Yes, This is how onActivityResult function looks like in the MainActivity.java file

<img width="942" alt="Screen Shot 2019-09-25 at 11 59 04" src="https://user-images.githubusercontent.com/33843864/65585922-f2e43100-df8b-11e9-8fce-45ccc8a6f3d2.png">

**Comment by pcova** [comment]
@vtolochk thank you. So much days spent on this

**Comment by jemise111** [comment]
Thanks @vtolochk! You saved me days of headache as well. One small note in case anyone else is using this:

`"com.braintreepayments.api.BraintreeFragment"` wasn't the correct tag to get the braintree fragment for me. I had to store the tag while creating the fragment and then use it later like this:

```
// inside Braintree.java

final BraintreeFragment mBraintreeFragment = BraintreeFragment.newInstance((AppCompatActivity)getCurrentActivity(), clientToken);
String braintreeFragmentTag = mBraintreeFragment.getTag();
MainActivity mainActivity = (MainActivity)getCurrentActivity();
mainActivity.setBraintreeFragmentTag(braintreeFragmentTag);
Venmo.authorizeAccount(mBraintreeFragment, false);

----------------------------------------------------------------

// inside MainActivity.java

public class MainActivity extends ReactActivity {
    private ReactInstanceManager mReactInstanceManager;
    private String mBraintreeFragmentTag;

    public void setBraintreeFragmentTag(String tag) {
        mBraintreeFragmentTag = tag;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        final int unmaskedRequestCode = requestCode & 0x0000ffff;
        Fragment braintreeFragment = getSupportFragmentManager().findFragmentByTag(mBraintreeFragmentTag);

        if (braintreeFragment != null) {
            braintreeFragment.onActivityResult(unmaskedRequestCode, resultCode, data);
        }
    }
```

**Comment by vtolochk** [comment]
I get this issue https://github.com/braintree/braintree_android/issues/312 for HostnameVerifier in RN, so I need to update Braintree to the > 3.14 version.

For those who use a solution that I posted earlier, I have updated the implementation to make it work on the newer version.

I am not sure about reliability, but for now, I don't have much choice.

![Screenshot 2020-12-01 at 11 12 09](https://user-images.githubusercontent.com/33843864/100720113-33c4aa80-33c6-11eb-91b4-05782fbf4178.png)
